### DISCUSSION TITLE
Java API

### DISCUSSION THREAD

**Discussion by egli** [daisy/pipeline#772] [Q&A]
The [documentation mentions](http://daisy.github.io/pipeline/Get-Help/API-Documentation/) that the "DAISY Pipeline 2 may also be used directly as a Java library instead of via client-server communication"

I though this approach might simplify my setup as I do not need to deal with installing the pipeline and creating jobs on the client side.

I tried to do this in a newly created project with the following code:
``` java
package ch.sbs.dp2;

import java.util.ServiceLoader;

import org.daisy.pipeline.script.ScriptRegistry;
import org.daisy.pipeline.script.ScriptService;

public class App 
{
    private static ServiceLoader<ScriptRegistry> scriptRegistryLoader = ServiceLoader.load(ScriptRegistry.class);

    public static void main( String[] args )
    {
	System.out.println( "Hello World!" );
	String scriptName = "dtbook-to-html";
	for (ScriptRegistry registry: scriptRegistryLoader) {
	    ScriptService script = registry.getScript(scriptName);
	    if (script != null) {
		System.out.println("Found script");
	    }
	}
    }
}
```
The pom.xml has a dependency on `framework-core`
``` xml
    <dependency>
      <groupId>org.daisy.pipeline</groupId>
      <artifactId>framework-core</artifactId>
      <version>8.0.0</version>
    </dependency>
```
when I run this with `mvn -B compile exec:java -Dexec.mainClass="ch.sbs.dp2.App"` I get an error saying 

```
[INFO] --- exec-maven-plugin:3.2.0:java (default-cli) @ daisy-pipeline-direct-client ---
Hello World!
[WARNING] 
java.util.ServiceConfigurationError: org.daisy.pipeline.script.ScriptRegistry: Provider org.daisy.pipeline.script.ScriptRegistry_SPI not found
    at java.util.ServiceLoader.fail (ServiceLoader.java:593)
    at java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass (ServiceLoader.java:1219)
    at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService (ServiceLoader.java:1228)
    at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext (ServiceLoader.java:1273)
    at java.util.ServiceLoader$2.hasNext (ServiceLoader.java:1309)
    at java.util.ServiceLoader$3.hasNext (ServiceLoader.java:1393)
    at ch.sbs.dp2.App.main (App.java:20)
    at org.codehaus.mojo.exec.ExecJavaMojo.doMain (ExecJavaMojo.java:385)
    at org.codehaus.mojo.exec.ExecJavaMojo.doExec (ExecJavaMojo.java:374)
    at org.codehaus.mojo.exec.ExecJavaMojo.lambda$execute$0 (ExecJavaMojo.java:296)
    at java.lang.Thread.run (Thread.java:833)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```
Does anyone have a complete running example on how to use the Pipeline as a library? Is this worth it in the first place? Can I include the pipeline in the Uberjar that I distribute?

Thanks
Christian

**Comment by bertfrees** [comment]
I don't know why org.daisy.pipeline.script.ScriptRegistry_SPI is not found. That is unexpected. If you want I can hava a look if you share the project with me.

By the way, including only `framework-core` will not be sufficient to do anything useful. You need at least few other framework jars, and the `dtbook-to-html` module for the conversion logic.

What I do in an application (that is not public) that uses Pipeline as a library is put this in my dependencies:

    <dependency>
      <groupId>org.daisy.pipeline</groupId>
      <artifactId>assembly</artifactId>
      <version>1.14.17-p1</version>
    </dependency>

This will simply pull in everything ([everything listed in the `dependencies` of this pom](https://repo1.maven.org/maven2/org/daisy/pipeline/assembly/1.14.17-p1/assembly-1.14.17-p1.pom)).

Including the Pipeline in an uberjar has not been tested yet.

**Comment by egli** [comment]
Hi @bertfrees thanks for the answer. Using the assembly dependency looks much better but unfortunately fails with [an error](https://github.com/sbsdev/daisy-pipeline-library-client/actions/runs/8613551056/job/23605068444).

I made the project public at https://github.com/sbsdev/daisy-pipeline-library-client

**Reply by bertfrees** [reply]
Oops, it looks like I haven't deployed these assembly artifacts on Maven Central yet. I will fix it and let you know when it's available.

**Reply by bertfrees** [reply]
@egli I've released the artifact. It should appear on Maven Central soon.
